### PR TITLE
Add docs build job

### PR DESCRIPTION
**Pull request by AyodeAwe**
The PR adds a `docs_build` process to the PR and Build workflows for this repository. The generated docs are synced to s3 for only the build workflows.

### CODE REVIEW

**Comment by BradReesWork** [comment]
@AyodeAwe there is a merge conflict

**Review by BradReesWork — APPROVED** [review]

File: ci/build_docs.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+pushd docs/cugraph
+sphinx-build -b dirhtml source _html
+sphinx-build -b text source _text
+popd
+
+
+if [[ "${RAPIDS_BUILD_TYPE}" == "branch" ]]; then
+  rapids-logger "Upload Docs to S3"
+  aws s3 sync --delete docs/cugraph/_html "s3://rapidsai-docs/cugraph/${VERSION_NUMBER}/html"
+  aws s3 sync --delete docs/cugraph/_text "s3://rapidsai-docs/cugraph/${VERSION_NUMBER}/text"
```
Comment (ajschmidt8) [inline ci/build_docs.sh:58]:
```suggestion
  aws s3 sync --delete docs/cugraph/_text "s3://rapidsai-docs/cugraph/${VERSION_NUMBER}/txt"
```

This S3 path should be consistent with the path used in the other PRs.

e.g. https://github.com/rapidsai/rmm/blob/7662b2411684079c58a1dc3d43d625b4a683f93f/ci/build_docs.sh#L44

File: compare.yaml [hunk 2]
```diff
@@ -0,0 +1,191 @@
+# Dependency list for https://github.com/rapidsai/dependency-file-generator
```
Comment (ajschmidt8) [inline compare.yaml:1]: This file looks like it was accidentally committed.

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

**Comment by AyodeAwe** [comment]
```
Handler <function process_generate_options at 0x7f26ae8f3ee0> for event 'builder-inited' threw an exception (exception: no module named cugraph_dgl)
```
I get this error locally as well.


`rapids-mamba-retry install "${PYTHON_CHANNEL}/linux-64/cugraph-dgl-*.tar.bz2"`
This bypass (which circumvents the `cudatoolkit` version mismatch) doesn't seem to work anymore for some reason. An interesting find while debugging locally is the following error when I try to import the cugraph_dgl module in a python script:

```
>>> import cugraph_dgl
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph_dgl/__init__.py", line 20, in <module>
    import cugraph_dgl.nn
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph_dgl/nn/__init__.py", line 13, in <module>
    from .conv import *  # noqa
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph_dgl/nn/conv/__init__.py", line 13, in <module>
    from .gatconv import GATConv
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph_dgl/nn/conv/gatconv.py", line 27, in <module>
    class GATConv(nn.Module):
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph_dgl/nn/conv/gatconv.py", line 120, in GATConv
    g: dgl.DGLHeteroGraph,
  File "/opt/conda/envs/docs/lib/python3.8/site-packages/cugraph/utilities/utils.py", line 403, in __getattr__
    raise RuntimeError(f"This feature requires the {self.name} " "package/module")
RuntimeError: This feature requires the dgl package/module
```

**Comment by AyodeAwe** [comment]
Regarding this [comment](https://github.com/rapidsai/cugraph/pull/3157#issuecomment-1407716999), a `pip install dgl` gets rid of that error but a new error pops up:

```
Extension error (sphinx.ext.autosummary):
Handler <function process_generate_options at 0x7f723bdb7ee0> for event 'builder-inited' threw an exception (exception: no module named cugraph_pyg.loader.dispatch)
```

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -52,6 +52,17 @@ jobs:
       branch: ${{ inputs.branch }}
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
+  docs-build:
+    if: ${{ startsWith(github.ref, 'refs/heads/branch-') }}
+    needs: python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/build.yaml:59]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

File: .github/workflows/pr.yaml [hunk 4]
```diff
@@ -62,6 +63,16 @@ jobs:
       arch: "amd64"
       container_image: "rapidsai/ci:latest"
       run_script: "ci/test_notebooks.sh"
+  docs-build:
+    needs: conda-python-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@cuda-118
```
Comment (ajschmidt8) [inline .github/workflows/pr.yaml:69]:
```suggestion
    uses: rapidsai/shared-action-workflows/.github/workflows/custom-job.yaml@branch-23.04
```

File: ci/build_docs.sh [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# This command installs `cugraph-dgl` without its dependencies
+# since this package can currently only run in `11.6` CTK environments
+# due to the dependency version specifications in its conda recipe.
+rapids-logger "Install cugraph-dgl"
+set -x
+rapids-mamba-retry install "${PYTHON_CHANNEL}/linux-64/cugraph-dgl-*.tar.bz2"
+set +x
+
+
+# Build Doxygen docs
```
Comment (ajschmidt8) [inline ci/build_docs.sh:44]:
```suggestion
```

We can remove this comment since it doesn't provide any value.

File: ci/build_docs.sh [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+rapids-logger "Build Doxygen docs"
+pushd cpp/doxygen
+doxygen Doxyfile
+popd
+
+# Build Python docs
```
Comment (ajschmidt8) [inline ci/build_docs.sh:50]:
```suggestion
```

We can remove this comment since it doesn't provide any value.

File: dependencies.yaml [hunk 7]
```diff
@@ -113,6 +119,22 @@ dependencies:
...
+      - output_types: [conda]
+        packages:
+          - doxygen
+          - graphviz
+          - ipython
+          - nbsphinx
+          - numpydoc
+          - pydata-sphinx-theme
+          - recommonmark
+          - sphinx_rtd_theme
```
Comment (ajschmidt8) [inline dependencies.yaml:133]:
```suggestion
```

We know from the [contents of conf.py](https://github.com/AyodeAwe/cugraph/blob/10c37b3932a5b1bda871bcabaa6d3cf2daf27d95/docs/cugraph/source/conf.py#L108) that `cugraph` uses the `pydata-sphinx-theme`.

Therefore we don't also need to install the `sphinx_rtd_theme`.

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

**Review by ajschmidt8 — APPROVED** [review]

**Comment by AyodeAwe** [comment]
@VibhuJawa how should this style check failure be handled? The error showed up after we imported [your patch](https://github.com/rapidsai/cugraph/pull/3157/commits/31bf9558ef5e56fdae1dee275f45f9a33806e5fe):

```
python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py:144:38: F821 undefined name 'create_tensorized_dataset'
```

https://github.com/rapidsai/cugraph/actions/runs/4247522209/jobs/7385666170

**Comment by VibhuJawa** [comment]
> @VibhuJawa how should this style check failure be handled? The error showed up after we imported [your patch](https://github.com/rapidsai/cugraph/pull/3157/commits/31bf9558ef5e56fdae1dee275f45f9a33806e5fe):
> 
> ```
> python/cugraph-dgl/cugraph_dgl/dataloading/dataloader.py:144:38: F821 undefined name 'create_tensorized_dataset'
> ```
> 
> https://github.com/rapidsai/cugraph/actions/runs/4247522209/jobs/7385666170
> 
> cc: @ajschmidt8

This  [line](https://github.com/rapidsai/cugraph/pull/3157/files#diff-90e104d897a4c013ed2a0277850b6ca609e86605dccf38da479b6a066c5e45feR144)  `create_tensorized_dataset` should have been `dgl.dataloading.create_tensorized_dataset`  like in this [patch](https://github.com/rapidsai/cugraph/pull/3157/commits/31bf9558ef5e56fdae1dee275f45f9a33806e5fe#diff-90e104d897a4c013ed2a0277850b6ca609e86605dccf38da479b6a066c5e45feR144) . .

I think we had a bad merge which erased it. Anyways, add it and you should be good to go.

**Review by BradReesWork — APPROVED** [review]

**Review by rlratzel — APPROVED** [review]
I see that @VibhuJawa looked over some of the non-docs related changes to cugraph-dgl files, so I'm assuming those are supposed to be in this PR.

**Comment by ajschmidt8** [comment]
> I see that @VibhuJawa looked over some of the non-docs related changes to cugraph-dgl files, so I'm assuming those are supposed to be in this PR.

Yes. Good call out, @rlratzel. There were some import issues that required a patch from Vibhu which we included.

**Review by VibhuJawa — APPROVED** [review]
LGTM

**Comment by VibhuJawa** [comment]
@rlratzel , Yup those changes allow building `cugraph-dgl` in a environment without  `dgl` and `pytorch` . The changes look good to me.  Approved the PR too.

**Pull request merged**